### PR TITLE
Scale subresource and Heapster querying in HorizontalPodAutoscaler

### PR DESCRIPTION
**Pull request by mwielgus**

### CODE REVIEW

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go [hunk 1]
```diff
@@ -18,23 +18,36 @@ package autoscalercontroller
...
+)
+
+var resourceToMetric = map[api.ResourceName]string{
+	api.ResourceCPU: "cpu-usage",
+}
+var heapsterQueryStart, _ = time.ParseDuration("-20m")
+
 type HorizontalPodAutoscalerController struct {
-	kubeClient unversioned.ExperimentalInterface
+	client    *client.Client
```
Comment (jszczepkowski) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: client should be also passed by Interface

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go [hunk 2]
```diff
@@ -48,11 +61,57 @@ func (a *HorizontalPodAutoscalerController) Run(syncPeriod time.Duration) {
...
+	list, err := a.expClient.HorizontalPodAutoscalers(ns).List(labels.Everything(), fields.Everything())
 	if err != nil {
 		return fmt.Errorf("error listing nodes: %v", err)
 	}
-	// TODO: implement!
-	glog.Info("autoscalers: %v", list)
+	for _, hpa := range list.Items {
+		reference := fmt.Sprintf("%s/%s/%s", hpa.Spec.ScaleRef.Kind, hpa.Spec.ScaleRef.Namespace, hpa.Spec.ScaleRef.Name)
+
+		scale, err := a.expClient.Scales(hpa.Spec.ScaleRef.Namespace).Get("replicationControllers", hpa.Spec.ScaleRef.Name)
```
Comment (jszczepkowski) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: "replicationControllers" -> hpa.Spec.ScaleRef.Kind

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/expapi"
+	//	endptspkg "k8s.io/kubernetes/pkg/api/endpoints"
```
Comment (jszczepkowski) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go]: Remove?

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	_ "k8s.io/kubernetes/pkg/api/latest"
+	"k8s.io/kubernetes/pkg/api/testapi"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	//	"k8s.io/kubernetes/pkg/client/unversioned/cache"
```
Comment (jszczepkowski) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go]: Remove?

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 5]
```diff
@@ -268,12 +269,12 @@ func (s *CMServer) Run(_ []string) error {
...
-	// 	glog.Fatalf("Invalid API configuration: %v", err)
-	// }
-	// horizontalPodAutoscalerController := autoscalercontroller.New(expClient)
-	// horizontalPodAutoscalerController.Run(s.NodeSyncPeriod)
+	expClient, err := client.NewExperimental(kubeconfig)
+	if err != nil {
+		glog.Fatalf("Invalid API configuration: %v", err)
+	}
+	horizontalPodAutoscalerController := autoscalercontroller.New(kubeClient, expClient)
+	horizontalPodAutoscalerController.Run(s.NodeSyncPeriod)
```
Comment (jszczepkowski) [inline cmd/kube-controller-manager/app/controllermanager.go]: Un-commenting & enabling this code by default is not a good idea. Please note that experimental API is disable by default, and our controller will spam logs with errors. We should leave it commented, or somehow associate it with turning on the experimental API.

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go (hunk 1 above)
Comment (mwielgus) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: I would love to but at this moment Interface doesn't support calling service via proxy (or I don't know how to do it).

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go (hunk 2 above)
Comment (mwielgus) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: sure, fixed.

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+	testServer, handlers := makeTestServer(t, hpaResponse, scaleResponse, podListResponse, heapsterResponse)
+	defer testServer.Close()
+	kubeClient := client.NewOrDie(&client.Config{Host: testServer.URL, Version: testapi.Version()})
+	expClient := client.NewExperimentalOrDie(&client.Config{Host: testServer.URL, Version: testapi.Version()})
+
+	hpaController := New(kubeClient, expClient)
+	err := hpaController.reconcileAutoscalers()
+
+	glog.Warningf("Error: %#v", err)
```
Comment (jszczepkowski) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go]: Check err != null?

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go (hunk 3 above)
Comment (mwielgus) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go]: done

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go (hunk 4 above)
Comment (mwielgus) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go]: done

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go (hunk 1 above)
Comment (jszczepkowski) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: OK. Just add a TODO, we will clean it up in a subsequent PR.

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go (hunk 6 above)
Comment (mwielgus) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller_test.go]: Done.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 5 above)
Comment (mwielgus) [inline cmd/kube-controller-manager/app/controllermanager.go]: Reverted.

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go (hunk 1 above)
Comment (mwielgus) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: Added.

**Comment by jszczepkowski** [comment]
LGTM

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test

**Comment by jszczepkowski** [comment]
Part of #12087.

File: pkg/controller/autoscaler/horizontalpodautoscaler_controller.go [hunk 7]
```diff
@@ -48,11 +62,58 @@ func (a *HorizontalPodAutoscalerController) Run(syncPeriod time.Duration) {
...
+			podNames = append(podNames, pod.Name)
+		}
+
+		metric, metricDefined := resourceToMetric[hpa.Spec.Target.Resource]
+		if !metricDefined {
+			glog.Warningf("Heapster metric not defined for %s %v", reference, hpa.Spec.Target.Resource)
+			continue
+		}
+		startTime := time.Now().Add(heapsterQueryStart)
+		metricPath := fmt.Sprintf("/api/v1/model/namespaces/%s/pod-list/%s/metrics/%s",
```
Comment (vishh) [inline pkg/controller/autoscaler/horizontalpodautoscaler_controller.go]: nit: We should switch to using client libraries for these APIs.

File: pkg/client/unversioned/scale.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+	return &scales{
+		client: c,
+		ns:     namespace,
+	}
+}
+
+// Get takes the reference to scale subresource and returns the subresource or error, if one occurs.
+func (c *scales) Get(kind string, name string) (result *expapi.Scale, err error) {
+	result = &expapi.Scale{}
+	if strings.ToLower(kind) == "replicationcontroller" {
```
Comment (jszczepkowski) [inline pkg/client/unversioned/scale.go]: This convertion is rather ugly.

@bgrant0607 what it the canonical way of converting kind ("replicationcontroller") into path ("replicationcontrollers")?

Comment (bgrant0607) [inline pkg/client/unversioned/scale.go]: https://github.com/kubernetes/kubernetes/blob/1c7ebc4a62f6f5df9d08a7ec4ce97835c769392e/pkg/api/meta/restmapper.go#L116

Comment (bgrant0607) [inline pkg/client/unversioned/scale.go]: Or more properly:
https://github.com/kubernetes/kubernetes/blob/1c7ebc4a62f6f5df9d08a7ec4ce97835c769392e/pkg/api/meta/restmapper.go#L162

**Pull request merged**